{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "807d2e79-7cb1-4bd5-978b-7eb58cc45e57",
   "metadata": {},
   "source": [
    "# Design HA inserts\n",
    "These will be added into our custom onboarded HA expression vector with Twist. In the Bloom lab, this plasmid has the ID# 2851. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cdb2e8bb-9a48-495c-97a6-f968a993d304",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import relevant packages\n",
    "import os\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import random\n",
    "\n",
    "\n",
    "# ID input and output\n",
    "datadir = '../data'\n",
    "resultsdir = '../results'\n",
    "\n",
    "ordersheetsdir = os.path.join(resultsdir, 'ordersheets')\n",
    "\n",
    "os.makedirs(datadir, exist_ok=True)\n",
    "os.makedirs(resultsdir, exist_ok=True)\n",
    "os.makedirs(ordersheetsdir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "538f8c6e-9df0-4abf-b069-04f0749f7018",
   "metadata": {},
   "source": [
    "There is a fixed upstream WSN packaging signal (19 amino acids) that we need to append the HA ectodomain for each of our sequences.\n",
    "\n",
    "The HA endodomain is kept constant for H3s and H1s.\n",
    "* For H3s, this represents amino acids 521-560 (when numbering from methionine start codon as 1)\n",
    "* For H1s, Andrea uses WSN sequence from amino acids 521 to end (position 568)\n",
    "\n",
    "Downstream, we need to add the mutated WSN packaging signal and a 16 nucleotide barcode. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "17d426b3-0e1e-43cc-bf7a-8116448830b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# First 19 amino acids of WSN\n",
    "wsn_upstream_19aa = 'atgaaggcaaaactactggtcctgttatatgcatttgtagctacagatgcagacaca'\n",
    "\n",
    "# First 20 amino acids of WSN\n",
    "wsn_upstream_20aa = 'atgaaggcaaaactactggtcctgttatatgcatttgtagctacagatgcagacacaata'\n",
    "\n",
    "# H1 endodomain, representing amino acids 521 to 568 (48 amino acids)\n",
    "h1_last_46aa_from_WSN = 'aaattggaatcaatgggagtgtatcagattctggcgatatattctacagtggcaagctccttagtactgctagtttctttaggagcgattagcttttggatgtgctccaacggCtcCCtAcaAtgTCgGatTtgTatTTAATAG'\n",
    "\n",
    "# H3 endodomain, representing amino acids 521 to 560 (40 amino acids)\n",
    "h3_endodomain = 'atcaagggagttgagctgaagtcaggatacaaagattggatcctatggatttcctttgccATGtcTtgCttCCtActGtgCgtAgcACtACtAggCttTatTatgtgggcGtgTcaGaaA'\n",
    "# Downstream mutated WSN packaging signal with double stop codon (11 amino acids)\n",
    "wsn_downstream = 'ggCtcCCtAcaAtgTCgGatTtgTatTTAATAG'\n",
    "\n",
    "# Barcode\n",
    "barcode = 'NNNNNNNNNNNNNNNN'\n",
    "nucleotides = ['a', 'c', 'g', 't']\n",
    "\n",
    "# Initialize barcode index (list of barcodes that have been used by a construct already)\n",
    "barcode_index = []\n",
    "\n",
    "# Add barcodes already used by prior viral libraries\n",
    "kikawa2023 = pd.read_csv(os.path.join(datadir, 'viral_libraries/2023_H3N2_Kikawa.csv'))['barcode'].tolist()\n",
    "loes2023 = pd.read_csv(os.path.join(datadir, 'viral_libraries/pdmH1N1_lib2023_loes.csv'))['barcode'].tolist()\n",
    "barcode_index.extend(loes2023)\n",
    "barcode_index.extend(kikawa2023)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb066ee9-3ace-41c4-bce2-91b16da64bf2",
   "metadata": {},
   "source": [
    "Define a function for designing inserts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1dad9378-c0ed-4aea-8aa7-9ce96b918612",
   "metadata": {},
   "outputs": [],
   "source": [
    "def design_inserts(subtype, \n",
    "                   insert_filepath, \n",
    "                   library_nucleotide_sequences,\n",
    "                   upstream_signalpep,\n",
    "                   ectodomain_start,\n",
    "                   ectodomain_length,\n",
    "                   endodomain_sequence,\n",
    "                   start_codon = \"ATGAAG\",\n",
    "                   append_additional_upstream_sequence = '',\n",
    "                   append_additional_downstream_sequence = '',\n",
    "                   virus_id = 1 # start virus naming at this index\n",
    "                  ):\n",
    "\n",
    "    # Only design if the ordersheet hasn't been generated\n",
    "    if os.path.exists(insert_filepath):\n",
    "        print(f\"File '{insert_filepath}' exists, reading that file and NOT regenerating barcodes.\")\n",
    "    else:\n",
    "        # Input FASTA file of subtype nucleotide sequences\n",
    "        fasta_file = library_nucleotide_sequences\n",
    "        \n",
    "        # Define the custom start codon to search for\n",
    "        start_codon = start_codon\n",
    "    \n",
    "        # Define ordersheet name parameters\n",
    "        virus_id = virus_id\n",
    "    \n",
    "        # Initialize empty ordersheet to populate with name, sequence\n",
    "        inserts = []\n",
    "    \n",
    "        # Set the number of barcodes to design for\n",
    "        n_barcodes = 2\n",
    "        \n",
    "        # Open FASTA file and design constructs for each entry\n",
    "        with open(fasta_file, \"r\") as handle:\n",
    "        \n",
    "            for record in SeqIO.parse(handle, \"fasta\"):\n",
    "                # Initialize barcode counter\n",
    "                i=1\n",
    "                \n",
    "                # Each strain needs barcodes designed \n",
    "                for n in list(range(0,n_barcodes)):\n",
    "        \n",
    "                    # Find the position of the first instance of 'ATGAAG'\n",
    "                    start_position = record.seq.find(start_codon)                \n",
    "                    assert start_position == -1, f\"For {record.id} - no start codon {start_codon} found\"\n",
    "\n",
    "                    # Extract the sequence starting from the found position \n",
    "                    insert_start = start_position + ectodomain_start # Insert will start after first 19 amino acids of WSN\n",
    "                    insert_end = start_position + (ectodomain_length) # Insert stops at amino acid 521\n",
    "                    ectodomain_insert_seq = record.seq[insert_start:insert_end]\n",
    "                    ectodomain_insert_translated_seq = ectodomain_insert_seq.translate()                \n",
    "                    # Identify the endodomain region (subtype specific)\n",
    "                    endodomain = endodomain_sequence\n",
    "                                    \n",
    "                    # Generate a barcode\n",
    "                    for n in list(range(0,100)): # Try 100 times to make a barcode \n",
    "                        barcode = ''.join(random.choices(nucleotides, k=16))\n",
    "                        if barcode[0:2] == 'gg': # Don't use barcodes that start with GG\n",
    "                            continue\n",
    "                        if barcode in barcode_index: # Don't use barcodes that have already been used in the library\n",
    "                            continue\n",
    "                        if n == 100:\n",
    "                            print('something really rare happened, try resetting barcode_index')\n",
    "                        else:\n",
    "                            barcode_index.append(barcode)\n",
    "                            break\n",
    "            \n",
    "                    # Expected sequence, including fixed upstream WSN signal peptide\n",
    "                    expected_seq = upstream_signalpep + ectodomain_insert_seq + endodomain + barcode\n",
    "                    # Insert sequence we need to order\n",
    "                    # Ectodomain, endodomain, and barcode\n",
    "                    insert_seq = ectodomain_insert_seq + endodomain + barcode\n",
    "\n",
    "                    # Add upstream sequence\n",
    "                    if append_additional_upstream_sequence == '':\n",
    "                        pass\n",
    "                    else:\n",
    "                        insert_seq = append_additional_upstream_sequence + insert_seq\n",
    "                    # Add downstream sequence\n",
    "                    if append_additional_downstream_sequence == '':\n",
    "                        pass\n",
    "                    else:\n",
    "                        insert_seq = insert_seq + append_additional_downstream_sequence\n",
    "            \n",
    "                    # Make a strain name with barcode info\n",
    "                    name = record.id\n",
    "                    name_barcoded = f'{subtype}_{virus_id}_bc{i}'\n",
    "                    i+=1\n",
    "\n",
    "                    # Get Genbank ID (and additional mutations) from FASTA header\n",
    "                    genbank_id = record.description[len(record.id):].strip('protein identical to ')\n",
    "    \n",
    "                    # Add to inserts list\n",
    "                    inserts.append([name, genbank_id, name_barcoded, str(insert_seq)])     \n",
    "\n",
    "                # Add to virus counter\n",
    "                virus_id+=1\n",
    "    \n",
    "            inserts_df = pd.DataFrame(inserts, columns = ['strain', 'genbank', 'name', 'sequence'])\n",
    "            # inserts_df = inserts_df.sort_values(by = 'name').reset_index(drop=True).to_csv(insert_filepath, index=False)\n",
    "            inserts_df = inserts_df.to_csv(insert_filepath, index=False)\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc2cd662-73e6-4cd5-b7aa-ae05365fc6f1",
   "metadata": {},
   "source": [
    "## Design H1 inserts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "08f6be17-4f20-4f2e-a709-5c67e0c6353c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File '../results/ordersheets/h1_inserts.csv' exists, reading that file and NOT regenerating barcodes.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>genbank</th>\n",
       "      <th>name</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A/Maryland/64/2024_H1N1</td>\n",
       "      <td>PV283376</td>\n",
       "      <td>H1N1_1_bc1</td>\n",
       "      <td>catttgtagctacagatgcagacacaataTGTATAGGTTATCATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A/Maryland/64/2024_H1N1</td>\n",
       "      <td>PV283376</td>\n",
       "      <td>H1N1_1_bc2</td>\n",
       "      <td>catttgtagctacagatgcagacacaataTGTATAGGTTATCATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A/Qinghai-Chengzhong/SWL1410/2024_H1N1</td>\n",
       "      <td>PQ850248</td>\n",
       "      <td>H1N1_2_bc1</td>\n",
       "      <td>catttgtagctacagatgcagacacaataTGTATAGGTTATCATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A/Qinghai-Chengzhong/SWL1410/2024_H1N1</td>\n",
       "      <td>PQ850248</td>\n",
       "      <td>H1N1_2_bc2</td>\n",
       "      <td>catttgtagctacagatgcagacacaataTGTATAGGTTATCATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A/Ulsan/492/2025_H1N1</td>\n",
       "      <td>PV100011</td>\n",
       "      <td>H1N1_3_bc1</td>\n",
       "      <td>catttgtagctacagatgcagacacaataTGTATAGGTTATCATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>A/Utah/39/2025_H1N1</td>\n",
       "      <td>PV550589</td>\n",
       "      <td>H1N1_36_bc2</td>\n",
       "      <td>catttgtagctacagatgcagacacaataTGTATAGGTTATCATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>A/Hawaii/ISC-1140/2025_H1N1</td>\n",
       "      <td>PV586141</td>\n",
       "      <td>H1N1_37_bc1</td>\n",
       "      <td>catttgtagctacagatgcagacacaataTGTATAGGTTATCATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>A/Hawaii/ISC-1140/2025_H1N1</td>\n",
       "      <td>PV586141</td>\n",
       "      <td>H1N1_37_bc2</td>\n",
       "      <td>catttgtagctacagatgcagacacaataTGTATAGGTTATCATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>A/NovaScotia/ET1801CP00018S/2025_H1N1</td>\n",
       "      <td>PV286420</td>\n",
       "      <td>H1N1_38_bc1</td>\n",
       "      <td>catttgtagctacagatgcagacacaataTGCATAGGTTATCATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>A/NovaScotia/ET1801CP00018S/2025_H1N1</td>\n",
       "      <td>PV286420</td>\n",
       "      <td>H1N1_38_bc2</td>\n",
       "      <td>catttgtagctacagatgcagacacaataTGCATAGGTTATCATGC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    strain   genbank         name  \\\n",
       "0                  A/Maryland/64/2024_H1N1  PV283376   H1N1_1_bc1   \n",
       "1                  A/Maryland/64/2024_H1N1  PV283376   H1N1_1_bc2   \n",
       "2   A/Qinghai-Chengzhong/SWL1410/2024_H1N1  PQ850248   H1N1_2_bc1   \n",
       "3   A/Qinghai-Chengzhong/SWL1410/2024_H1N1  PQ850248   H1N1_2_bc2   \n",
       "4                    A/Ulsan/492/2025_H1N1  PV100011   H1N1_3_bc1   \n",
       "..                                     ...       ...          ...   \n",
       "71                     A/Utah/39/2025_H1N1  PV550589  H1N1_36_bc2   \n",
       "72             A/Hawaii/ISC-1140/2025_H1N1  PV586141  H1N1_37_bc1   \n",
       "73             A/Hawaii/ISC-1140/2025_H1N1  PV586141  H1N1_37_bc2   \n",
       "74   A/NovaScotia/ET1801CP00018S/2025_H1N1  PV286420  H1N1_38_bc1   \n",
       "75   A/NovaScotia/ET1801CP00018S/2025_H1N1  PV286420  H1N1_38_bc2   \n",
       "\n",
       "                                             sequence  \n",
       "0   catttgtagctacagatgcagacacaataTGTATAGGTTATCATGC...  \n",
       "1   catttgtagctacagatgcagacacaataTGTATAGGTTATCATGC...  \n",
       "2   catttgtagctacagatgcagacacaataTGTATAGGTTATCATGC...  \n",
       "3   catttgtagctacagatgcagacacaataTGTATAGGTTATCATGC...  \n",
       "4   catttgtagctacagatgcagacacaataTGTATAGGTTATCATGC...  \n",
       "..                                                ...  \n",
       "71  catttgtagctacagatgcagacacaataTGTATAGGTTATCATGC...  \n",
       "72  catttgtagctacagatgcagacacaataTGTATAGGTTATCATGC...  \n",
       "73  catttgtagctacagatgcagacacaataTGTATAGGTTATCATGC...  \n",
       "74  catttgtagctacagatgcagacacaataTGCATAGGTTATCATGC...  \n",
       "75  catttgtagctacagatgcagacacaataTGCATAGGTTATCATGC...  \n",
       "\n",
       "[76 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "design_inserts(\n",
    "    subtype = 'H1N1',\n",
    "    insert_filepath = os.path.join(ordersheetsdir, 'h1_inserts.csv'),\n",
    "    library_nucleotide_sequences = '../results/strains_for_library/h1_nt_seqs_for_library.fasta',\n",
    "    upstream_signalpep = wsn_upstream_20aa,\n",
    "    ectodomain_start = 20*3,\n",
    "    ectodomain_length = 520*3,\n",
    "    endodomain_sequence = h1_last_46aa_from_WSN,\n",
    "    append_additional_upstream_sequence = 'catttgtagctacagatgcagacaca' + 'ata', # Overlap with WSN signal peptide\n",
    "    append_additional_downstream_sequence = 'AGATCGGAAGAGCGTCGTGT', # Overlap with Illumina R1 priming sequence             \n",
    ")\n",
    "\n",
    "h1_inserts_df = pd.read_csv(os.path.join(ordersheetsdir, 'h1_inserts.csv'))\n",
    "h1_inserts_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32b9ebdd-c467-4924-a76f-dd7c9dc6d518",
   "metadata": {},
   "source": [
    "Write FASTA files with the insert sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dd3d4596-f0d1-4a98-967b-edc3144aff99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize output directory\n",
    "output_dir = os.path.join(resultsdir, \"insert_fasta\")\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Write one FASTA file per row\n",
    "for _, row in h1_inserts_df.iterrows():\n",
    "    header = f\">{row['strain']}_{row['name']}_{row['genbank']}\"\n",
    "    sequence = row['sequence']\n",
    "    fasta_content = f\"{header}\\n{sequence}\\n\"\n",
    "    \n",
    "    # Define file name \n",
    "    filename = f\"{row['strain'].replace('/','-').strip('_H1N1')}_{row['name']}_{row['genbank']}.fasta\"\n",
    "    filepath = os.path.join(output_dir, filename)\n",
    "    \n",
    "    with open(filepath, \"w\") as f:\n",
    "        f.write(fasta_content)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b008c371-7608-4ad8-ae92-c09edfa18afa",
   "metadata": {},
   "source": [
    "## Design H3 inserts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d2d3f3d6-6ca0-4dc2-8cfc-eefe1915a0c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File '../results/ordersheets/h3_inserts.csv' exists, reading that file and NOT regenerating barcodes.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>genbank</th>\n",
       "      <th>name</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A/Washington/284/2024_H3N2</td>\n",
       "      <td>PV001009</td>\n",
       "      <td>H3N2_1_bc1</td>\n",
       "      <td>catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A/Washington/284/2024_H3N2</td>\n",
       "      <td>PV001009</td>\n",
       "      <td>H3N2_1_bc2</td>\n",
       "      <td>catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A/Wisconsin/NIRC-IS-1028/2024_H3N2</td>\n",
       "      <td>PV227971</td>\n",
       "      <td>H3N2_2_bc1</td>\n",
       "      <td>catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A/Wisconsin/NIRC-IS-1028/2024_H3N2</td>\n",
       "      <td>PV227971</td>\n",
       "      <td>H3N2_2_bc2</td>\n",
       "      <td>catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A/Maldives/2147/2024_H3N2</td>\n",
       "      <td>PV220008</td>\n",
       "      <td>H3N2_3_bc1</td>\n",
       "      <td>catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>A/Amapa/021563-IEC/2024_H3N2</td>\n",
       "      <td>PV507313</td>\n",
       "      <td>H3N2_74_bc2</td>\n",
       "      <td>catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>A/New_York/39/2025_H3N2</td>\n",
       "      <td>PV509521</td>\n",
       "      <td>H3N2_75_bc1</td>\n",
       "      <td>catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>A/New_York/39/2025_H3N2</td>\n",
       "      <td>PV509521</td>\n",
       "      <td>H3N2_75_bc2</td>\n",
       "      <td>catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>A/New_York/GKISBBBE61555/2025_H3N2</td>\n",
       "      <td>PV271403</td>\n",
       "      <td>H3N2_76_bc1</td>\n",
       "      <td>catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>A/New_York/GKISBBBE61555/2025_H3N2</td>\n",
       "      <td>PV271403</td>\n",
       "      <td>H3N2_76_bc2</td>\n",
       "      <td>catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>152 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 strain   genbank         name  \\\n",
       "0            A/Washington/284/2024_H3N2  PV001009   H3N2_1_bc1   \n",
       "1            A/Washington/284/2024_H3N2  PV001009   H3N2_1_bc2   \n",
       "2    A/Wisconsin/NIRC-IS-1028/2024_H3N2  PV227971   H3N2_2_bc1   \n",
       "3    A/Wisconsin/NIRC-IS-1028/2024_H3N2  PV227971   H3N2_2_bc2   \n",
       "4             A/Maldives/2147/2024_H3N2  PV220008   H3N2_3_bc1   \n",
       "..                                  ...       ...          ...   \n",
       "147        A/Amapa/021563-IEC/2024_H3N2  PV507313  H3N2_74_bc2   \n",
       "148             A/New_York/39/2025_H3N2  PV509521  H3N2_75_bc1   \n",
       "149             A/New_York/39/2025_H3N2  PV509521  H3N2_75_bc2   \n",
       "150  A/New_York/GKISBBBE61555/2025_H3N2  PV271403  H3N2_76_bc1   \n",
       "151  A/New_York/GKISBBBE61555/2025_H3N2  PV271403  H3N2_76_bc2   \n",
       "\n",
       "                                              sequence  \n",
       "0    catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...  \n",
       "1    catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...  \n",
       "2    catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...  \n",
       "3    catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...  \n",
       "4    catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...  \n",
       "..                                                 ...  \n",
       "147  catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...  \n",
       "148  catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...  \n",
       "149  catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...  \n",
       "150  catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...  \n",
       "151  catttgtagctacagatgcagacacaCAAAAAATACCTGGAAATGA...  \n",
       "\n",
       "[152 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "design_inserts(\n",
    "    subtype = 'H3N2',\n",
    "    insert_filepath = os.path.join(ordersheetsdir, 'h3_inserts.csv'),\n",
    "    library_nucleotide_sequences = '../results/strains_for_library/h3_nt_seqs_for_library.fasta',\n",
    "    upstream_signalpep = wsn_upstream_19aa,\n",
    "    ectodomain_start = 16*3,\n",
    "    ectodomain_length = 517*3,\n",
    "    endodomain_sequence = h3_endodomain + wsn_downstream,\n",
    "    append_additional_upstream_sequence = 'catttgtagctacagatgcagacaca', # Overlap with WSN signal peptide\n",
    "    append_additional_downstream_sequence = 'AGATCGGAAGAGCGTCGTGT', # Overlap with Illumina R1 priming sequence                    \n",
    ")\n",
    "\n",
    "h3_inserts_df = pd.read_csv(os.path.join(ordersheetsdir, 'h3_inserts.csv'))\n",
    "h3_inserts_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28dcec30-4095-470e-b694-c1246fa40938",
   "metadata": {},
   "source": [
    "Write FASTA files with the insert sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4d9fbeac-ae61-4fb6-b245-de27e5271b17",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize output directory\n",
    "output_dir = os.path.join(resultsdir, \"insert_fasta\")\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# Write one FASTA file per row\n",
    "for _, row in h3_inserts_df.iterrows():\n",
    "    header = f\">{row['strain']}_{row['name']}_{row['genbank']}\"\n",
    "    sequence = row['sequence']\n",
    "    fasta_content = f\"{header}\\n{sequence}\\n\"\n",
    "    \n",
    "    # Define file name \n",
    "    filename = f\"{row['strain'].replace('/','-').strip('_H3N2').strip('_(H3N2)')}_{row['name']}_{row['genbank']}.fasta\"\n",
    "    filepath = os.path.join(output_dir, filename)\n",
    "    \n",
    "    with open(filepath, \"w\") as f:\n",
    "        f.write(fasta_content)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9edbe144-9cb9-42c5-9eb9-c4f4c8d2378b",
   "metadata": {},
   "source": [
    "## Write ordersheet for 2025-05-09"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0db65251-80f7-40bc-80f6-c61ba8fe7d6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "inserts_df = pd.concat([h1_inserts_df, h3_inserts_df])\n",
    "\n",
    "# Save ordersheet\n",
    "inserts_df[['name', 'sequence']].to_csv(os.path.join(ordersheetsdir, '2025-05-09_ordersheet.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cad9bc4-7205-49d0-9650-d4170c8e432e",
   "metadata": {},
   "source": [
    "## Design vaccine strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0b86198b-0680-4c46-ad4b-fbf7a68901fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File '../results/ordersheets/vaccine_inserts.csv' exists, reading that file and NOT regenerating barcodes.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>genbank</th>\n",
       "      <th>name</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A/Croatia/10136RV/2023_H3N2</td>\n",
       "      <td>PV262858_D202A</td>\n",
       "      <td>H3N2_77_bc1</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaatacctggaaatga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A/Croatia/10136RV/2023_H3N2</td>\n",
       "      <td>PV262858_D202A</td>\n",
       "      <td>H3N2_77_bc2</td>\n",
       "      <td>catttgtagctacagatgcagacacacaaaaaatacctggaaatga...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        strain         genbank         name  \\\n",
       "0  A/Croatia/10136RV/2023_H3N2  PV262858_D202A  H3N2_77_bc1   \n",
       "1  A/Croatia/10136RV/2023_H3N2  PV262858_D202A  H3N2_77_bc2   \n",
       "\n",
       "                                            sequence  \n",
       "0  catttgtagctacagatgcagacacacaaaaaatacctggaaatga...  \n",
       "1  catttgtagctacagatgcagacacacaaaaaatacctggaaatga...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "design_inserts(\n",
    "    subtype = 'H3N2',\n",
    "    insert_filepath = os.path.join(ordersheetsdir, 'vaccine_inserts.csv'),\n",
    "    library_nucleotide_sequences = '../data/sequences/2025-05-16_vaccine_sequences.fasta',\n",
    "    upstream_signalpep = wsn_upstream_19aa,\n",
    "    ectodomain_start = 16*3+1,\n",
    "    ectodomain_length = 517*3+1,\n",
    "    endodomain_sequence = h3_endodomain + wsn_downstream,\n",
    "    append_additional_upstream_sequence = 'catttgtagctacagatgcagacaca', # Overlap with WSN signal peptide\n",
    "    append_additional_downstream_sequence = 'AGATCGGAAGAGCGTCGTGT', # Overlap with Illumina R1 priming sequence\n",
    "    virus_id = 77\n",
    ")\n",
    "\n",
    "vaccine_inserts = pd.read_csv(os.path.join(ordersheetsdir, 'vaccine_inserts.csv'))\n",
    "vaccine_inserts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e120d057-95cd-4dab-88bc-a5a154f9952c",
   "metadata": {},
   "source": [
    "## Write ordersheet for 2025-05-16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "61b51130-2093-4462-a905-59d961dfaffb",
   "metadata": {},
   "outputs": [],
   "source": [
    "inserts_df = pd.concat([vaccine_inserts])\n",
    "\n",
    "# Save ordersheet\n",
    "inserts_df[['name', 'sequence']].to_csv(os.path.join(ordersheetsdir, '2025-05-16_ordersheet.csv'), index=False)\n",
    "\n",
    "## Manually added a Lisboa/188 strain to this list!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba7f1c17-9a21-4c9d-9259-edbd6aa38935",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a51bce64-5116-4ff9-aac3-1a521220fe36",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
